### PR TITLE
Implement "pure C wrapper" entry point into minimal C++ engine

### LINKED ISSUE #7: Implement pure C wrapper entry point into minimum C++ BindEngine
Engine will be essentially a "minimum useless engine" at this point, but this sets up the C++ structure that we want.

- [x] Set up the C++ / pure C wrapper entry point
- [x] Get `BindEngine` and `InitEngine` to a "minimum useless engine" state

### PR DESCRIPTION
**Pull request by ZacharyEspiritu**
Resolves #7.

This PR initializes the engine codebase with the implementation of the "pure C wrapper" entry point into a "minimum useless C++ engine".

### Main Changes

- Sets up the C++ / pure C wrapper entry point from OpenSSL into `EngineBind`
- Implements `EngineBind` such that engine is in a "minimum useless engine" state
- Adds unit tests to make sure expected fields of initialized `ENGINE` struct are set

### Questions

- In the design doc, there's a `libengine` top-level namespace (with the intent that libkms has its own `libkms` top-level namespace). Is it preferable to have a single top-level namespace for the whole project though (seems like that's what the style guide suggests)?
- I put `/test` as a separate directory from `/src` since I was anticipating that it might be good to group all of the tests (including the future integration tests) under one folder. However, this requires a `default_visibility` directive in the `BUILD` file in `/src` to expose `/src` targets to `/test`. Is it more preferable to put the unit tests in the `/src` directory with the code they test? (I see that [grpc](https://github.com/grpc/grpc) separates src and test, while [tensorflow](https://github.com/tensorflow/tensorflow) keeps unit tests in the same directory.)
- For future reference, is this is a good length for PRs?

### CODE REVIEW

**Comment by ajfeldma-google** [comment]
> Resolves #7.
> 
> This PR initializes the engine codebase with the implementation of the "pure C wrapper" entry point into a "minimum useless C++ engine".
> 
> ### Main Changes
> * Sets up the C++ / pure C wrapper entry point from OpenSSL into `EngineBind`
> * Implements `EngineBind` such that engine is in a "minimum useless engine" state
> * Adds unit tests to make sure expected fields of initialized `ENGINE` struct are set
> 
> ### Questions
> * In the design doc, there's a `libengine` top-level namespace (with the intent that libkms has its own `libkms` top-level namespace). Is it preferable to have a single top-level namespace for the whole project though (seems like that's what the style guide suggests)?

Often there is a top level namespace and then multiple 2nd level namespaces, and so I could imagine a top-level namespace for the whole project and then a different 2nd level namespace for each library.

> * I put `/test` as a separate directory from `/src` since I was anticipating that it might be good to group all of the tests (including the future integration tests) under one folder. However, this requires a `default_visibility` directive in the `BUILD` file in `/src` to expose `/src` targets to `/test`. Is it more preferable to put the unit tests in the `/src` directory with the code they test? (I see that [grpc](https://github.com/grpc/grpc) separates src and test, while [tensorflow](https://github.com/tensorflow/tensorflow) keeps unit tests in the same directory.)

Internally, the convention is to put the unit tests in the same directory as the source code files they test. e.g.,
```
./foo.cc
./foo_test.cc
```
> * For future reference, is this is a good length for PRs?

This length is great, but not every PR description has to be this detailed.

File: WORKSPACE [hunk 1]
```diff
@@ -15,15 +15,39 @@
 
 load("@bazel_tools//tools/build_defs/repo:git.bzl", "git_repository")
 
+"""
+Google Cloud C++ API dependencies.
+"""
+
+# Import Google APIs with C++ rules.
 git_repository(
-    name = "gtest",
```
Comment (ajfeldma-google) [inline WORKSPACE:19]: Presumably you'll need `gtest` and `abseil-cpp` soon, and so why remove them?

File: test/libengine/engine_bind_test.cc [hunk 2]
```diff
@@ -0,0 +1,67 @@
+/**
```
Comment (ajfeldma-google) [inline test/libengine/engine_bind_test.cc]: As I mentioned in my PR comment, it's more idiomatic for Google to have tests and src in the same directory,

File: src/libengine/engine_bind.cc [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+#include "src/libengine/engine_bind.h"
+
+#include <openssl/engine.h>
+
+#include "src/libengine/engine_name.h"
+#include "src/libengine/engine_setup.h"
+
+namespace libengine {
```
Comment (ajfeldma-google) [inline src/libengine/engine_bind.cc]: I'd leave off the `lib`: `engine` instead of `libengine`

**Review by ajfeldma-google — CHANGES_REQUESTED** [review]

File: WORKSPACE (hunk 1 above)
Comment (ZacharyEspiritu) [reply WORKSPACE:19]: These are imported through the com_github_googleapis_... deps already.

https://github.com/googleinterns/cloud-kms-oss-tools/blob/5f7d08435890a98a5a705103a93b12427f503e62/WORKSPACE#L30-L35

File: src/libengine/engine_bind.cc (hunk 3 above)
Comment (ZacharyEspiritu) [reply src/libengine/engine_bind.cc]: I added a `kmsengine` top-level namespace and renamed `namespace libengine` to `namespace engine`, though not sure if there'll be confusion between `engine` and `kmsengine`.

Comment (ZacharyEspiritu) [reply src/libengine/engine_bind.cc]: If it's more mnemonic as to how the code flows, we could call the two .so namespaces `frontend` and `backend`, or something else that sacrifices the "engine" naming for a more directional flavor. Not sure if it really matters though, but probably easier to rename things now before more files get added.

File: src/libengine/BUILD [hunk 4]
```diff
@@ -14,10 +14,10 @@
 # limitations under the License.
 #
 
-package(default_visibility = ["//test:__subpackages__"])
```
Comment (tonyliaoss) [inline src/libengine/BUILD]: May be easier using an autoformatter for changes like these, but I haven't looked into it too much.

e.g. https://github.com/bazelbuild/buildtools

Probably not worth it to worry about this now if it's a lot of effort though. We can always add something like this when we're done implementing the main functionalities.

File: src/libengine/engine_bind.cc (hunk 3 above)
Comment (tonyliaoss) [reply src/libengine/engine_bind.cc]: What's the difference between `kmsengine` and `engine`? 😅

`::kmsengine::engine::DoFoo()` seems a bit awkward?

File: src/engine/engine_bind_test.cc [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  void TearDown() override {
+    ENGINE_free(engine);
+  }
+
+  ENGINE *engine;
+};
+
+TEST_F(EngineBindTest, SetsEngineId) {
```
Comment (tonyliaoss) [inline src/engine/engine_bind_test.cc]: If all you do is engine_bind (i.e. the test body is not substantially different from each other), maybe just combine all the verifications into one?

File: src/libengine/engine_bind.cc (hunk 3 above)
Comment (ZacharyEspiritu) [reply src/libengine/engine_bind.cc]: See discussion in this thread above - we discussed in the design-doc adding a top-level namespace shared between both libengine and libkms, and then in this thread we decided to rename `namespace libengine` to `namespace engine`. I wasn't sure if there would be confusion between `engine` and `kmsengine`, so maybe another name to replace `engine` is in order.

File: src/libengine/BUILD (hunk 4 above)
Comment (ZacharyEspiritu) [reply src/libengine/BUILD]: I've been running `buildifier` when I change Bazel files, but I've only been doing it manually from the command line. I can look into adding a `genrule` that just runs the `buildifier` command on all of the Bazel files when needed.

File: src/engine/engine_bind_test.cc (hunk 5 above)
Comment (ZacharyEspiritu) [reply src/engine/engine_bind_test.cc]: Do you mean something like this?

```c++
TEST_F(EngineBindTest, InitializesExpectedEngineStructFields) {
  EngineBind(engine, NULL);
  EXPECT_STREQ(ENGINE_get_id(engine), kEngineId);
  EXPECT_STREQ(ENGINE_get_name(engine), kEngineName);
  EXPECT_NE(ENGINE_get_init_function(engine), nullptr);
  EXPECT_NE(ENGINE_get_finish_function(engine), nullptr);
  EXPECT_TRUE(ENGINE_get_flags(engine) & ENGINE_FLAGS_NO_REGISTER_ALL);
}
```

Comment (ZacharyEspiritu) [reply src/engine/engine_bind_test.cc]: In this case, is it still worth it to have the text fixture that calls `ENGINE_new` and `ENGINE_free`?

Comment (tonyliaoss) [reply src/engine/engine_bind_test.cc]: Yeah something like that. As for text fixture, that's up to you. Do you foresee other checks being added later on to this file that would require test fixtures?

Comment (ZacharyEspiritu) [reply src/engine/engine_bind_test.cc]: At the moment, I can't think of anything that can't just be added as another `EXPECT` so I'll just remove the fixture for now.

File: WORKSPACE (hunk 1 above)
Comment (ajfeldma-google) [reply WORKSPACE:19]: Fair enough. It might be worth adding a comment explaining this.

File: src/libengine/engine_bind.cc (hunk 3 above)
Comment (ajfeldma-google) [reply src/libengine/engine_bind.cc]: You could rename `engine` `openssl_interface` or something.

**Review by ajfeldma-google — COMMENTED** [review]
LGTM other than the namespace name issue and adding a comment that grpc includes gtest and absl

**Review by ajfeldma-google — DISMISSED** [review]

**Comment by ZacharyEspiritu** [comment]
libengine namespace is now `namespace bridge`.

**Review by tonyliaoss — APPROVED** [review]

**Review by ajfeldma-google — APPROVED** [review]

**Pull request merged**